### PR TITLE
DZ5: Тестирование API (requests + json schema)

### PR DESCRIPTION
**Pull request by Mikhail-QA-O**

### CODE REVIEW

File: api_testing/test_getting_resource.py [hunk 1]
```diff
@@ -0,0 +1,12 @@
+from schemas.data_preparation import ValidationJsonSchemas
+
+
+def test_get_user_1(session, base_url):
+    get_user_1 = session.get(base_url + '1')
+    assert get_user_1.status_code == 200
+    ValidationJsonSchemas.checking_json_schema(url=get_user_1)
```
Comment (agridyaev) [inline api_testing/test_getting_resource.py]: You could add test case for todo with max id (**200**) to complete boundary value analysis. Also you could add check API response for correct todo id:

```
@pytest.mark.parametrize('todo_id', [1, 200])
def test_get_todo(session, base_url, todo_id):
    res = session.get(base_url + str(todo_id))
    assert res.status_code == 200
    assert res.json()['id'] == todo_id
    ValidationJsonSchemas.checking_json_schema(url=res)
```

File: api_testing/test_getting_resource.py [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+
+
+def test_get_invalid_data(session, base_url):
+    get_invalid_data = session.get(base_url + '0')
+    assert get_invalid_data.status_code == 404
```
Comment (agridyaev) [inline api_testing/test_getting_resource.py]: You could add **-1** and **201** todos ids to complete boundary value analysis. Also you could add check length of API response:

```
@pytest.mark.parametrize('todo_id', [-1, 0, 201])
def test_get_invalid_data(session, base_url, todo_id):
    res = session.get(base_url + str(todo_id))
    assert res.status_code == 404
    assert len(res.json()) == 0
```

File: api_testing/test_listing_resource.py [hunk 3]
```diff
@@ -0,0 +1,7 @@
+from schemas.data_preparation import ValidationJsonSchemas
+
+
+def test_get_all_data(session, base_url):
+    get_all_data = session.get(base_url)
+    assert get_all_data.status_code == 200
+    ValidationJsonSchemas.checking_json_schema_array(url=get_all_data)
```
Comment (agridyaev) [inline api_testing/test_listing_resource.py]: You could add a check for the length of todos in API response:

```
def test_get_all_data(session, base_url):
    get_all_data = session.get(base_url)
    assert get_all_data.status_code == 200
    assert len(get_all_data.json()) == 200
    ValidationJsonSchemas.checking_json_schema_array(url=get_all_data)
```

File: api_testing/test_creating_resource.py [hunk 4]
```diff
@@ -0,0 +1,11 @@
+from api_testing.credentials import CredentialsCreating
+
+
+def test_creating_resource(session, base_url):
+    creating_resource = session.post(base_url, json=CredentialsCreating.valid_data)
+    assert creating_resource.status_code == 201
```
Comment (agridyaev) [inline api_testing/test_creating_resource.py]: You could add checks for **userId**, **id**, **title** and **completed** response fields:

```
def test_creating_resource(session, base_url):
    res = session.post(base_url, json=CredentialsCreating.valid_data)
    assert res.status_code == 201
    r_json = res.json()
    assert r_json['userId'] == CredentialsCreating.valid_data['userId']
    assert r_json['id'] == CredentialsCreating.valid_data['id']
    assert r_json['title'] == CredentialsCreating.valid_data['title']
    assert r_json['completed'] == CredentialsCreating.valid_data['completed']
```

File: api_testing/credentials.py [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+        "title": "199",
+        "completed": True,
+    }
+
+    user_201 = {
+        "userId": 199,
+        "id": 199,
+        "title": "199",
+        "completed": True,
+    }
```
Comment (agridyaev) [inline api_testing/credentials.py]: Please check the values for **user_201**, I guess the should be equal to:

```
user_201 = {
        "userId": 201,
        "id": 201,
        "title": "201",
        "completed": True,
    }
```

File: api_testing/test_updating_resource.py [hunk 6]
```diff
@@ -0,0 +1,18 @@
+from api_testing.credentials import CredentialUpdating
+
+
+def test_update_user_199(session, base_url):
+    update_user_10 = session.put(base_url + '199', json=CredentialUpdating.user_199)
+    assert update_user_10.status_code == 200
+    assert dict(update_user_10.json()).get('id') == 199
```
Comment (agridyaev) [inline api_testing/test_updating_resource.py]: You could add checks for **userId**, **id**, **title** and **completed** response fields:

```
def test_update_user_199(session, base_url):
    res = session.put(base_url + '199', json=CredentialUpdating.user_199)
    assert res.status_code == 200
    r_json = res.json()
    assert r_json['userId'] == CredentialUpdating.user_199['userId']
    assert r_json['id'] == CredentialUpdating.user_199['id']
    assert r_json['title'] == CredentialUpdating.user_199['title']
    assert r_json['completed'] == CredentialUpdating.user_199['completed']
```

File: api_testing/test_updating_resource.py [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+
+def test_update_user_201(session, base_url):
+    update_user_201 = session.put(base_url + '201', json=CredentialUpdating.user_201)
+    assert update_user_201.status_code == 500
+
+
+def test_update_title_user_10(session, base_url):
+    update_title = session.patch(base_url + '10', json=CredentialUpdating.new_title)
+    assert update_title.status_code == 200
+    assert dict(update_title.json()).get('title') == '123'
```
Comment (agridyaev) [inline api_testing/test_updating_resource.py]: You could add tests for updating all todo fields - **userId**, **id**, **title** and **completed**:

```
@pytest.mark.parametrize('field, value', [
    ('id', 11),
    ('userId', 10),
    ('title', '123'),
    ('completed', False),
    ('completed', True),
])
def test_update_title_user_10(session, base_url, field, value):
    res = session.patch(base_url + '10', json={field: value})
    assert res.status_code == 200
    assert res.json()[field] == value
```

File: api_testing/test_deleting_resource.py [hunk 8]
```diff
@@ -0,0 +1,4 @@
+def test_remove_user_2(session, base_url):
+    delete_user_2 = session.delete(base_url + '2')
+    assert delete_user_2.status_code == 200
+    assert dict(delete_user_2.json()).get('id') is None
```
Comment (agridyaev) [inline api_testing/test_deleting_resource.py]: You could simplify this check:

```
def test_remove_user_2(session, base_url):
    res = session.delete(base_url + '2')
    assert res.status_code == 200
    assert not res.json()
```

File: api_testing/test_filtering_resource.py [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+    pytest.param(
+        ('?completed=NONE'), [],
+        id='completed'
+    ),
+
+])
+def test_filter_invalid_value_key(session, base_url, wrong_value, result):
+    filter_invalid_value_key = session.get(base_url + wrong_value)
+    assert filter_invalid_value_key.status_code == 200
+    assert filter_invalid_value_key.json() == result
```
Comment (agridyaev) [inline api_testing/test_filtering_resource.py]: You could simplify this test:

```
@pytest.mark.parametrize('wrong_value', [
    '?userId=300',
    '?id=300',
    '?title=otus',
    '?completed=NONE'
])
def test_filter_invalid_value_key(session, base_url, wrong_value):
    res = session.get(base_url + wrong_value)
    assert res.status_code == 200
    assert not res.json()
```

File: api_testing/test_filtering_resource.py [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+
+
+def test_filter_user_id_1(session, base_url):
+    filter_user_id_1 = session.get(base_url, params=CredentialsFiltering.user_id_1)
+    assert filter_user_id_1.status_code == 200
```
Comment (agridyaev) [inline api_testing/test_filtering_resource.py]: You could add tests for filtering by **userId**, **id**, **title** and **completed** fields:

```
@pytest.mark.parametrize('field, value', [
    ('userId', 1),
    ('id', 1),
    ('title', 'delectus aut autem'),
    ('completed', False),
    ('completed', True)
])
def test_filter_positive(session, base_url, field, value):
    str_val = str(value).lower() if isinstance(value, bool) else str(value)
    res = session.get(base_url, params={field: str_val})
    assert res.status_code == 200
    j_res = res.json()
    assert len(j_res) > 0
    for x in j_res:
        assert x[field] == value
```

File: schemas/data_preparation.py [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+class ValidationJsonSchemas:
+    @staticmethod
+    def open_json_schema(data, schema_file):
+        with open(schema_file) as f:
+            read_json_schema = json.load(f)
+        return validate(instance=data, schema=read_json_schema)
+
+    @staticmethod
+    def checking_json_schema(url):
+        ValidationJsonSchemas.open_json_schema(url.json(), '/study/schemas/schema.json')
```
Comment (agridyaev) [inline schemas/data_preparation.py]: You could use relative path here:

```
ValidationJsonSchemas.open_json_schema(url.json(), 'schemas/schema.json')
```

File: schemas/schema_array.json [hunk 12]
```diff
@@ -0,0 +1,6 @@
+{
+  "type": "array",
+  "items": {
+    "$ref": "file:/study/schemas/schema.json"
```
Comment (agridyaev) [inline schemas/schema_array.json]: You could use relative path here:

```
"$ref": "file:schemas/schema.json"
```
